pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -395,6 +422,92 @@ func listEqualUnordered(a []string, b []string) bool {
 	return true
 }
 
+// handleAck will detect if the message is an ACK or NACK, and update/log/count
+// using the generic structures. "Classical" CDS/LDS/RDS/EDS use separate logic -
+// this is used for the API-based LDS and generic messages.
+func (s *DiscoveryServer) handleAck(con *XdsConnection, discReq *xdsapi.DiscoveryRequest) bool {
+	// Handle watching Istio config types. This provides similar functionality with MCP.
+	// Alternative to using 8080:/debug/configz
+	// Names are based on the current stable naming in istiod.
+	if discReq.ErrorDetail != nil {
+		errCode := codes.Code(discReq.ErrorDetail.Code)
+		adsLog.Warnf("ADS: ACK ERROR %s %s:%s", con.ConID, errCode.String(), discReq.ErrorDetail.GetMessage())
+		return true
+	}
+
+	t := discReq.TypeUrl
+	if discReq.ResponseNonce != "" {
+		con.mu.RLock()
+		routeNonceSent := con.NonceSent[t]
+		routeVersionInfoSent := con.RouteVersionInfoSent
+		con.mu.RUnlock()
+		if routeNonceSent != "" && routeNonceSent != discReq.ResponseNonce {
+			adsLog.Debugf("ADS:RDS: Expired nonce received %s, sent %s, received %s",
+				con.ConID, routeNonceSent, discReq.ResponseNonce)
+			rdsExpiredNonce.Increment()
+			return true
+		}
+		// GRPC doesn't send version info in NACKs for RDS. Technically if nonce matches
+		// previous response, it is an ACK/NACK.
+		if discReq.VersionInfo == routeVersionInfoSent ||
+				routeNonceSent == discReq.ResponseNonce	{
+			adsLog.Debugf("ADS: ACK %s %s %s", con.ConID, discReq.VersionInfo, discReq.ResponseNonce)
+			con.mu.Lock()
+			con.RouteNonceAcked = discReq.ResponseNonce
+			con.mu.Unlock()
+		}
+		return true
+	}
+
+	return false
+}
+
+// Handle a custom config resource
+func (s *DiscoveryServer) handleResource(con *XdsConnection, discReq *xdsapi.DiscoveryRequest) error {
```

Comment:
I can split it to separate PR - in reality both gRPC and this change are tied to moving DNS resolution to the agent. The gRPC ApiListener conveys name -> cluster VIP mappings. The ServiceEntry resource allows resolution of external names. 

Instead of using a newly invented gRPC service ( or using CoreDNS one ) - I would rather use the current XDS protocol. And instead of adding another one-off code for ServiceEntry - it seems more useful to have a generic code that can return all collections.